src/fsharp/FSharp.Core/prim-types.fs

```diff
@@ -4465,31 +4465,33 @@ namespace Microsoft.FSharp.Core
 
              // Using 'let x = (# ... #) in x.ToString()' leads to better IL, without it, an extra stloc and ldloca.s (get address-of)
              // gets emitted, which are unnecessary. With it, the extra address-of-variable is not created
-             when 'T : float      = let x = (# "" value : float #) in x.ToString(null, CultureInfo.InvariantCulture)
-             when 'T : float32    = let x = (# "" value : float32 #) in x.ToString(null, CultureInfo.InvariantCulture)
-             when 'T : decimal    = let x = (# "" value : decimal #) in x.ToString(null, CultureInfo.InvariantCulture)
+             when 'T : float      = let x = (# "" value : float #)      in x.ToString(null, CultureInfo.InvariantCulture)
+             when 'T : float32    = let x = (# "" value : float32 #)    in x.ToString(null, CultureInfo.InvariantCulture)
+             when 'T : decimal    = let x = (# "" value : decimal #)    in x.ToString(null, CultureInfo.InvariantCulture)
              when 'T : BigInteger = let x = (# "" value : BigInteger #) in x.ToString(null, CultureInfo.InvariantCulture)
              
              // no IFormattable
-             when 'T : char       = let x = (# "" value : char #) in x.ToString()
-             when 'T : bool       = let x = (# "" value : bool #) in x.ToString()
+             when 'T : char       = let x = (# "" value : char #)       in x.ToString()
+             when 'T : bool       = let x = (# "" value : bool #)       in x.ToString()
+             when 'T : nativeint  = let x = (# "" value : nativeint #)  in x.ToString()
+             when 'T : unativeint = let x = (# "" value : unativeint #) in x.ToString()
 
              // For the int-types:
              // It is not possible to distinguish statically between Enum and (any type of) int.
              // This way we'll print their symbolic value, as opposed to their integral one 
              // E.g.: 'string ConsoleKey.Backspace' gives "Backspace", rather than "8")
              when 'T : sbyte      = (box value :?> IFormattable).ToString(null, CultureInfo.InvariantCulture)
```

Comment:
Can't we use the `constrained !!T.` opcode with the `(# ... #)` syntax for the signed integers as well to avoid boxing too? Will it bypass the type system?

Reply:
@teo-tsirpanis, it'll break, because I cannot use `ToString()`, instead I must use `IFormattable::ToString(...)`. For the `IFormattable` call, to bypass boxing, you need to emit the `call` on the exact `type + method`. Since I don't know the type, I cannot choose between, say `System.Int32::ToString(null, InvariantCulture)` and `System.ConsoleKey::ToString(null, InvariantCulture)`. The only alternative is to use `callvirt`. Adding `constrained` doesn't help here, because then it would always resolve to the integer call, instead of enum OR integer.

I've asked several people, but so far nobody has a clue as to how to fix improve this further.

I've updated the original text of this PR with new timings. As you can see there (`UL` type vs `L` type), it certainly would help performance if we could emit such calls.